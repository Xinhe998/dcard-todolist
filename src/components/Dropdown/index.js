import React, { useContext, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Store from '../../reducers/context';
import * as action from '../../actions/todoActions';
import useClickOutside from '../../hooks/useClickOutside';

import './index.scss';

const Dropdown = ({
  icon,
  text,
  options,
  swichOptionHandler,
  isOpen,
  handleClickDispatch,
}) => {
  const dispatch = useContext(Store);
  const [selectedOption, setSelectedOption] = useState(options[0]);
  const dropdownRef = useRef();
  useClickOutside(isOpen, dropdownRef, () => swichOptionHandler(false));
  return (
    <div className="Dropdown" ref={dropdownRef}>
      <button
        type="button"
        className="Dropdown__button"
        onClick={() => swichOptionHandler(!isOpen)}
      >
        <img className="Dropdown__button__icon" src={icon} />
        {text}
        {selectedOption}
      </button>
      {isOpen ? (
        <div className="Dropdown__tooltip">
          <ul className="Dropdown__tooltip__list">
            {options.map((option, index) => (
              <li
                key={index}
                onClick={() => {
                  setSelectedOption(option);
                  swichOptionHandler(false);
                  handleClickDispatch(option);
                }}
                className={selectedOption === option ? 'active' : ''}
              >
                {option}
              </li>
            ))}
          </ul>
        </div>
      ) : null}
    </div>
  );
};

Dropdown.propTypes = {
  icon: PropTypes.string,
  text: PropTypes.string,
  options: PropTypes.array,
  swichOptionHandler: PropTypes.func,
  isOpen: PropTypes.bool,
  handleClickDispatch: PropTypes.func,
};

Dropdown.defaultProps = {
  icon: '',
  text: '',
  options: [],
  swichOptionHandler: null,
  isOpen: false,
  handleClickDispatch: null,
};
export default Dropdown;
