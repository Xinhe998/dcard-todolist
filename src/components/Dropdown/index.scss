.Dropdown {
  height: 40px;
  border-radius: 25px;
  background: rgba(255, 255, 255, 0.17);
  user-select: none;
  color: #ffffff;
  margin-bottom: 20px;
  transition: all 0.5s;
  position: relative;
  &:hover {
    box-shadow: 0px 3px 6px rgba(0, 0, 0, 0.16);
  }
}
.Dropdown__button {
  cursor: pointer;
  width: 100%;
  height: 100%;
  border: none;
  background: none;
  color: #ffffff;
  font-size: 14px;
  font-weight: 500;
  outline: 0;
  padding: 0 20px;
  text-align: left;
  .Dropdown__button__icon {
    width: 20px;
    height: 20px;
    vertical-align: middle;
    margin-right: 10px;
  }
}
.Dropdown__tooltip {
  width: 100%;
  border-radius: 10px;
  background: #ffffff;
  box-shadow: var(--card-box-shadow);
  color: var(--text-color);
  position: absolute;
  padding: 0;
  margin-top: 10px;
  text-align: center;
  z-index: 100;
  &::before {
    content: '';
    width: 0;
    height: 0;
    border-style: solid;
    border-width: 0 10px 10px 10px;
    border-color: transparent transparent #ffffff transparent;
    position: absolute;
    top: -8px;
    left: calc((100% - 10px) / 2);
  }
  .Dropdown__tooltip__list {
    list-style: none;
    padding: 0;
    li {
      cursor: pointer;
      padding: 10px 30px;
      &:hover, &.active {
        background: var(--secondary-color-gray2);
      }
    }
  }
}
