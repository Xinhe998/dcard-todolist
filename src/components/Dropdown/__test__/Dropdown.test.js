/* eslint-disable func-names */
import 'babel-polyfill';
import React from 'react';
import { shallow, configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Dropdown from '../index';

import filter from '../../../assets/filter.png';

// 以該解析器提供給 Enzyme 做渲染 Component 的設置
configure({ adapter: new Adapter() });

describe('<Dropdown />', function () {
  let testComp;
  beforeEach(() => {
    jest.resetModules();
    this.params = {
      icon: null,
      text: '',
      options: [],
      swichOptionHandler: jest.fn(),
      isOpen: true,
      handleClickDispatch: jest.fn(),
    };

    this.makeSubject = () => {
      const {
        icon,
        text,
        options,
        swichOptionHandler,
        isOpen,
        handleClickDispatch,
      } = this.params;

      testComp = shallow(
        <Dropdown
          icon={icon}
          text={text}
          options={options}
          swichOptionHandler={swichOptionHandler}
          isOpen={isOpen}
          handleClickDispatch={handleClickDispatch}
        />,
      );
      return testComp;
    };
  });
  describe('when it has complete data', () => {
    beforeEach(() => {
      this.params.icon = filter;
      this.params.text = 'Test:';
      this.params.options = ['option1', 'option2', 'option3'];
      this.subject = this.makeSubject();
    });
    it('should be same as snapshot', async () => {
      const wrapper = render(await testComp);
      expect(wrapper).toMatchSnapshot();
    });
    it('should be three options', async () => {
      expect(testComp.find('.Dropdown__tooltip__list__option')).toHaveLength(3);
    });
    // 測試icon
    it('should render icon', () => {
      expect(testComp.find('img').prop('src')).toEqual(filter);
    });
    // 測試text
    it('should render text', () => {
      expect(testComp.find('.Dropdown__button').text()).toEqual('Test:option1');
    });
    // 測試按下 button 會觸發 swichOptionHandler
    it('test click event', () => {
      testComp.find('button').simulate('click');
      expect(this.params.swichOptionHandler.mock.calls.length).toEqual(1);
    });
    // 測試按下選項
    it('test click event', () => {
      testComp.find('.Dropdown__tooltip__list__option').at(0).simulate('click');
      expect(this.params.swichOptionHandler.mock.calls.length).toEqual(1);
      expect(this.params.handleClickDispatch.mock.calls.length).toEqual(1);
    });
  });
});
