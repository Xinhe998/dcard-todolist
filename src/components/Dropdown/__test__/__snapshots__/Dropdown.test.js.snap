// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> when it has complete data should be same as snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "Dropdown",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "Dropdown__button",
          "type": "button",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "Dropdown__button__icon",
              "src": "filter.png",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "data": "Test:option1",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "text",
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "src": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "src": undefined,
            },
          },
          Object {
            "data": "Test:option1",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "Dropdown__button__icon",
                "src": "filter.png",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "src": undefined,
              },
            },
            "type": "text",
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "Dropdown__tooltip",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "Dropdown__tooltip__list",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option active",
                  },
                  "children": Array [
                    Object {
                      "data": "option1",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option",
                    },
                    "children": Array [
                      Object {
                        "data": "option2",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {
                        "class": "Dropdown__tooltip__list__option",
                      },
                      "children": Array [
                        Object {
                          "data": "option3",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "li",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option",
                  },
                  "children": Array [
                    Object {
                      "data": "option2",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option",
                    },
                    "children": Array [
                      Object {
                        "data": "option3",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option active",
                    },
                    "children": Array [
                      Object {
                        "data": "option1",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option",
                  },
                  "children": Array [
                    Object {
                      "data": "option3",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option",
                    },
                    "children": Array [
                      Object {
                        "data": "option2",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {
                        "class": "Dropdown__tooltip__list__option active",
                      },
                      "children": Array [
                        Object {
                          "data": "option1",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "li",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "ul",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "type": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "Dropdown__tooltip",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "Dropdown__tooltip__list",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "Dropdown__tooltip__list__option active",
                },
                "children": Array [
                  Object {
                    "data": "option1",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "li",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option",
                  },
                  "children": Array [
                    Object {
                      "data": "option2",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option",
                    },
                    "children": Array [
                      Object {
                        "data": "option3",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "Dropdown__tooltip__list__option",
                },
                "children": Array [
                  Object {
                    "data": "option2",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "li",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option",
                  },
                  "children": Array [
                    Object {
                      "data": "option3",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option active",
                  },
                  "children": Array [
                    Object {
                      "data": "option1",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "Dropdown__tooltip__list__option",
                },
                "children": Array [
                  Object {
                    "data": "option3",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "li",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "Dropdown__tooltip__list__option",
                  },
                  "children": Array [
                    Object {
                      "data": "option2",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "Dropdown__tooltip__list__option active",
                    },
                    "children": Array [
                      Object {
                        "data": "option1",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "ul",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "Dropdown__button",
            "type": "button",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "Dropdown__button__icon",
                "src": "filter.png",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "data": "Test:option1",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "text",
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "src": undefined,
              },
            },
            Object {
              "data": "Test:option1",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "Dropdown__button__icon",
                  "src": "filter.png",
                },
                "children": Array [],
                "name": "img",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "src": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "src": undefined,
                },
              },
              "type": "text",
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "type": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
